--- app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
+++ app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
@@ -5821,6 +5821,20 @@
     }
   }
 
+  @Test
+  fun testFlashback_featureFlagOff_thenFeatureFlagOn() {
+    // Start with feature flag disabled
+    executeInPreviousAppInstance { component ->
+      TestPlatformParameterModule.forceEnableFlashbackSupport(false)
+    }
+
+    // Enable feature flag and continue exploration
+    setUpTestWithFlashbackFeatureOn()
+    launchForExploration(TEST_EXPLORATION_ID_2, shouldSavePartialProgress = false).use {
+      startPlayingExploration()
+      moveToFlashbackState()
+    }
+  }
+
   @Test
   fun testFragment_flashbackFeatureOn_onSubmitWrongRatioAnswer_flashbackButtonIsVisible() {
