--- a/bazel/gen-deps.py
+++ b/bazel/gen-deps.py
@@ -17,6 +17,9 @@
         # e.g.
         # org.robolectric\:android-all-instrumented\:4.1.2_r1-robolectric-r1.jar=../../org_robolectric_android_all_instrumented_4_1_2_r1_robolectric_r1/jar/android-all-instrumented-4.1.2_r1-robolectric-r1.jar
         print("%s=%s" % (getKey(jar), getValue(jar)))
+        if "-i7" in getKey(jar):
+            print("%s=%s" % (getKey(jar).replace("-i7", "-i2"), getValue(jar)))
+            print("%s=%s" % (getKey(jar).replace("-i7", "-i4"), getValue(jar)))
 
 
 if __name__ == "__main__":