--- third_party/googleservices/BUILD
+++ third_party/googleservices/BUILD
@@ -1,3 +1,4 @@
+load("@rules_android//rules:rules.bzl", "android_library", "aar_import", "android_binary")
 java_binary(
   name = "GenerateGoogleServicesXml",
   srcs = ["java/com/google/gms/googleservices/GenerateGoogleServicesXml.java"],

--- tools/crashlytics/BUILD
+++ tools/crashlytics/BUILD
@@ -1,3 +1,4 @@
+load("@rules_android//rules:rules.bzl", "android_library", "aar_import", "android_binary")
 exports_files(["defs.bzl"])
 
 java_binary(

--- tools/crashlytics/defs.bzl
+++ tools/crashlytics/defs.bzl
@@ -1,3 +1,4 @@
+load("@rules_android//rules:rules.bzl", "android_library")
 def crashlytics_android_library(name, package_name, build_id, resource_files):
   _CRASHLYTICS_PROP_TEMPLATE = \
   """build_id={build_id}
@@ -61,7 +62,7 @@ package_name={package_name}"""
       cmd = "$(location @tools_android//tools/crashlytics) \"%s\" $@" % crashlytics_manifest_file_content,
   )
 
-  native.android_library(
+  android_library(
       name = name,
       assets = [crashlytics_properties_file],
       assets_dir = "_%s_crashlytics/assets" % name,

--- tools/googleservices/BUILD
+++ tools/googleservices/BUILD
@@ -1 +1,2 @@
+load("@rules_android//rules:rules.bzl", "android_library", "aar_import", "android_binary")
 exports_files(["defs.bzl"])

--- tools/googleservices/defs.bzl
+++ tools/googleservices/defs.bzl
@@ -1,1 +1,2 @@
+load("@rules_android//rules:rules.bzl", "android_library")
 """Macros to support Google services, e.g. Firebase Cloud Messaging."""
